lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -192,185 +214,337 @@ class AccessConflictAnalysis {
 
-  /// Tracks the in-scope accesses at the end of each block, for the purpose of
-  /// finding nested conflicts.  (Out-of-scope accesses are currently only
-  /// tracked locally for the purpose of merging access scopes.)
-  llvm::SmallDenseMap<SILBasicBlock *, DenseAccessSet, 32> blockOutAccess;
+  using RegionToRegionStorageMap =
+      llvm::SmallDenseMap<unsigned, RegionStorage *>;
+  RegionToRegionStorageMap regionToStorageMap;
+  AccessedStorageLocs storageAccesses;
 
   Result result;
-
 public:
-  AccessConflictAnalysis(SILFunction *F, PostOrderFunctionInfo *PO,
-                         AccessedStorageAnalysis *ASA)
-      : F(F), PO(PO), ASA(ASA) {}
+  RegionStorageAnalysis(LoopRegionFunctionInfo *LRFI,
+                        AccessedStorageAnalysis *ASA)
+      : LRFI(LRFI), ASA(ASA) {}
+  ~RegionStorageAnalysis();
 
   Result &&analyze() &&;
-
 protected:
   void identifyBeginAccesses();
+  void identifyStorageLocations();
+  void processBlockRegion(unsigned id);
+  void processLoopRegion(unsigned id);
 
-  void visitBeginAccess(BeginAccessInst *beginAccess,
-                        SparseAccessSet &accessMap);
+  void mergeBBPredAccesses(unsigned succID);
 
-  void visitEndAccess(EndAccessInst *endAccess, SparseAccessSet &accessMap);
+  void visitBeginAccess(BeginAccessInst *beginAccess);
+  void visitEndAccess(EndAccessInst *endAccess);
+  void detectApplyConflicts(const DenseAccessSet &conflictFreeSet,
+                            RegionStorage *applyStorage,
+                            const FunctionAccessedStorage &callSiteAccesses,
+                            const FullApplySite &fullApply);
 
-  void visitFullApply(FullApplySite fullApply, SparseAccessSet &accessMap);
+  void visitFullApply(FullApplySite fullApply);
 
-  void mergePredAccesses(SILBasicBlock *succBB,
-                         SparseAccessSet &mergedAccesses);
+  void calcBottomUpOrder(LoopRegion *region,
+                         llvm::SmallVectorImpl<unsigned> &worklist);
 
-  void visitBlock(SILBasicBlock *BB);
+  RegionStorage *getRegionStorageFromIns(SILInstruction *instr);
 };
 
-/// A sparse set of in-flight accesses.
-///
-/// Explodes a DenseAccessSet into a temporary sparse set for comparison
-/// and membership.
-///
-/// The AccessConflictAnalysis result provides the instruction to index
-/// mapping.
-class SparseAccessSet {
-  AccessConflictAnalysis::Result &result;
-
-  // Mark the in-scope accesses.
-  // (Most functions have < 64 accesses.)
-  llvm::SmallBitVector inScopeBitmask;
-  // Mark a potential conflicts on each access since the last begin/end marker.
-  llvm::SmallBitVector conflictBitmask;
-  DenseAccessSet denseVec; // Hold all local accesses seen thus far.
+/// The per-region result of RegionStorageAnalysis.
+class RegionStorage {
+  LoopRegion *R;
+  RegionStorageAnalysis::Result &result;
+
+  // A set of accesses we care about in this scope.
+  RegionStorageAnalysis::AccessedStorageLocs storageAccesses;
```

Comment:
`storageAccesses` leads me to think this is a set of accesses. Shouldn't it be 
`AccessedStorageSet accessedStorageSet`?